### PR TITLE
Fixed #24591 -- Optimized cloning of ModelState objects.

### PR DESCRIPTION
**Pull request by knbk**
Changed ModelState.clone() to create a shallow copy of self.fields
and self.managers.

https://code.djangoproject.com/ticket/24591

### CODE REVIEW

**Comment by claudep** [comment]
I think that a happy test suite is a good sign :-) Looks good to me.

**Comment by timgraham** [comment]
Are the commits meant to be squashed at all? You might add a docstring or more to `ImmutableWrapper`.

**Comment by charettes** [comment]
I have concerns about this patch. From my understanding fields are actually recreated to prevent them from keeping reference to previous states models. This patch works fine when dealing with basic fields since their only reference to a model is their `model` property assigned upon `contribute_to_class`.

Related fields on the other end can hold two additional reference to models: `rel.to` and `rel.through`. It looks like the current approach doesn't prevent `rel` alteration and could result in really hard to debug issues -- the main reason why we keep things simple by simply copying the fields.

**Comment by knbk** [comment]
Deconstruction in `from_model()` prevent that the fields contain any references to an actual model class - they are all converted to string references. An extra check in `__init__()` to ensure that they are string references is not a bad idea, though. 

If you want to prevent hard to debug issues, I think that on second thought I'd rather stick with deepcopy than with the current `ImmutableWrapper`. It is slower, but the impact if you do change the field (which you must not do) is smaller and more predictable. It is still _significantly_ faster than recreating the field. 

Field deconstruction is one of the most performance-critical part of migrations, right next to database IO, and it is incredibly slow. That, and people are starting to report rendering times close to an hour [1]. This patch is a massive improvement on cpu and memory usage. 

[1] https://code.djangoproject.com/ticket/24554

@timgraham I'm not entirely convinced that the `ImmutableWrapper` is the best approach in this patch. I'll squash the commits when I'm happy with the result.

**Comment by charettes** [comment]
Hi @knbk,

First thanks for working on this, I also agree that we should try solving this performance critical part of migrations.

I like the idea of a check in `__init__()` and the use of `deepcopy` and I'd be interested in learning how the performance enhancement compare to the actual solution your proposing.

I'm also wondering if performance is still an issue since (#24554)[https://code.djangoproject.com/ticket/24554]. From your comments this brings significant speed ups and memory saving but are migrations really that slow now that the previous fix has been committed?

I think I'm a bit reluctant here because I had to deal with previous state model reference bugs in the past and I also know that it was the case for @MarkusH and it was a bit of a headache for me.

**Comment by MarkusH** [comment]
I pretty much digged through the entire state handling again and am still alive. Can't be that bad then :grin:

As of now the fields are reconstructed each and every time a model or a relates model is changed twice: once in `ModelState.clone()` and once in `ModelState.render()`. This is the correct behavior, since the operations can possibly do anything with the model state they want.

As it turns out we, luckily, never modify the field instances inside of operations. For field related operations we replace the respective field instance if it changes with the one from the migration file. If only a field's name changes we reuse the field instance. This is still save because the instance doesn't know about its name. In model operations we take the time to actually clone the field instance where necessary: https://github.com/django/django/blob/master/django/db/migrations/operations/models.py#L183 . This also explains why there are no failing test cases with the `ImmutableWrapper` applied on the fields.

Bottom line, I think we are safe if we just use `self.fields` and put a big attention box in the docs stating that operations must noch change an existing field instance.

`ModelState.__init__()` does already some introspection of fields and checks that a field is not bound to a model. Adding something for relational fields doesn't sound too bad, though I'm unsure where the bound model should come from. The important thing here is that `ModelState.render()` reconstructs the field.

**Comment by charettes** [comment]
Thanks for your input Markus. If you're confident this can go as is I'll retract my concerns.

**Comment by MarkusH** [comment]
Your concerns are valid. We can speed things up here if and only if we make it clear to everybody working on built-in or 3rd party migration operations that field instances must not be changed. If we do that, I'm ok with just using `self.field`.

In that case we need to start talking about which version this patch is going to land in. It is a backwards incompatible change in a public/private? API, so it doesn't/does qualify for 1.8.x.

File: django/db/migrations/state.py [hunk 1]
```diff
@@ -304,12 +304,23 @@ def __init__(self, app_label, name, fields, options=None, bases=None, managers=N
...
             raise ValueError("ModelState.fields cannot be a dict - it must be a list of 2-tuples.")
-        # Sanity-check that fields are NOT already bound to a model.
         for name, field in fields:
+            # Sanity-check that fields are NOT already bound to a model.
             if hasattr(field, 'model'):
                 raise ValueError(
                     'ModelState.fields cannot be bound to a model - "%s" is.' % name
                 )
+            # Sanity-check that relation fields are NOT referring to a model class.
+            if field.remote_field and hasattr(field.remote_field.model, '_meta'):
```
Comment (MarkusH) [inline django/db/migrations/state.py]: Any reason why we can't go with `field.is_relation` and `field.related_model` here?

**Comment by MarkusH** [comment]
LGTM. @timgraham do you have any comments on the docs?

File: docs/ref/migration-operations.txt [hunk 2]
```diff
@@ -404,6 +404,13 @@ easy to read and cover a lot of the example usage of semi-internal aspects
 of the migration framework like ``ProjectState`` and the patterns used to get
 historical models.
 
+.. warning::
+
+    Due to performance reasons, the :class:`~django.db.models.Field`
```
Comment (timgraham) [inline docs/ref/migration-operations.txt]: Due to -> For

File: docs/ref/migration-operations.txt [hunk 3]
```diff
@@ -404,6 +404,13 @@ easy to read and cover a lot of the example usage of semi-internal aspects
...
+    instances in ``ModelState`` are reused across migrations. You must never
+    directly change the fields. To change a field, replace it with a new field
```
Comment (timgraham) [inline docs/ref/migration-operations.txt]: changes the fields how? I am not sure what this is warning is saying not to do.

File: docs/releases/1.9.txt [hunk 4]
```diff
@@ -360,6 +360,15 @@ Miscellaneous
 * ``CommaSeparatedIntegerField`` validation has been refined to forbid values
   like ``','``, ``',1'``, and ``'1,,2'``.
 
+* ``ModelState`` has been changed to check that the ``to`` and ``through``
```
Comment (timgraham) [inline docs/releases/1.9.txt]: I am not sure these release notes are necessary. At least, I think most users won't know what this means (I don't), and I don't know that we have broken any public APIs?

Comment (MarkusH) [inline docs/releases/1.9.txt]: Point taken. I thought I had already seen packages out there overriding the `state_forwards()` method of operations, although it's private API. But so far the Github search doesn't give me anything: https://github.com/search?utf8=%E2%9C%93&q=django+migrations+Operation+state_forwards&type=Code&ref=searchresults

File: docs/ref/migration-operations.txt [hunk 5]
```diff
@@ -402,7 +402,18 @@ You can take this template and work from it, though we suggest looking at the
...
-historical models.
+historical models, as well as ``ModelState`` and the patterns used to mutate
+historical models in ``state_forwards``.
+
+.. warning::
+
+    For performance reasons, the :class:`~django.db.models.Field` instances in
+    ``ModelState.fields`` are reused across migrations. You must never change
+    the attributes on these instances. If you need to mutate a field in
+    ``state_forwards``, you must remove the old instance from
```
Comment (timgraham) [inline docs/ref/migration-operations.txt]: state_forwards()

File: docs/ref/migration-operations.txt [hunk 6]
```diff
@@ -402,7 +402,18 @@ You can take this template and work from it, though we suggest looking at the
...
+    ``ModelState.fields``, and add a new instance in its place. The same is true
```
Comment (timgraham) [inline docs/ref/migration-operations.txt]: no comma needed

File: docs/ref/migration-operations.txt [hunk 7]
```diff
@@ -402,7 +402,18 @@ You can take this template and work from it, though we suggest looking at the
 built-in Django operations in ``django.db.migrations.operations`` - they're
 easy to read and cover a lot of the example usage of semi-internal aspects
 of the migration framework like ``ProjectState`` and the patterns used to get
-historical models.
+historical models, as well as ``ModelState`` and the patterns used to mutate
+historical models in ``state_forwards``.
```
Comment (timgraham) [inline docs/ref/migration-operations.txt]: () as below

**Comment by timgraham** [comment]
Thanks, those docs make sense to me now.

File: docs/ref/migration-operations.txt [hunk 8]
```diff
@@ -402,7 +402,18 @@ You can take this template and work from it, though we suggest looking at the
...
+
+.. warning::
```
Comment (timgraham) [inline docs/ref/migration-operations.txt]: Do you think this is a common issue? I am not sure a warning box is needed. The placing seems a bit odd (before what seem to be more basic instructions).

Comment (knbk) [inline docs/ref/migration-operations.txt]: I don't think it's a common issue, but it can be extremely difficult to debug, and it's not obvious if you're not very familiar with this part of Django. Custom operations are not needed for 99% of the cases (even the example should be using `RunSQL`), so I think a good fat warning box for the people who do end up in this part of the documentation and need to use Django's internals is not a bad idea. 

I didn't put a lot of thought in the placement, I guess I could move it more towards the end, after the list of notes.

**Comment by timgraham** [comment]
merged in 1a1f16d67da7080241ad9f6b4325f22364ac57ba, thanks!
